File: beware-of-the-sheep/src/World.cpp
#include "pch.h"
#include "World.h"
#include "Entities/Animals.h"
#include "Entities/Plants.h"


namespace bots {

	World::World(unsigned int width, unsigned int height) :
		m_width(width), m_height(height),
		m_area(width, height)
	{
		srand(static_cast<unsigned int>(time(nullptr)));

		spawnEntity(Entity::Kind::Human, { 1, 1 });
		//spawnEntity(Entity::Kind::CyberSheep, { 15, 15 });

		spawnEntity(Entity::Kind::Antelope, { 3, 3 });
		spawnEntity(Entity::Kind::Bellandona, { 5, 5 });
		spawnEntity(Entity::Kind::Dandelion, { 7, 7 });
		spawnEntity(Entity::Kind::Fox, { 9, 9 });
		spawnEntity(Entity::Kind::Grass, { 11, 11 });
		spawnEntity(Entity::Kind::Guarana, { 13, 13 });
		spawnEntity(Entity::Kind::Hogweed, { 15, 15 });
		spawnEntity(Entity::Kind::Sheep, { 17, 17 });
		spawnEntity(Entity::Kind::Turtle, { 19, 19 });
		spawnEntity(Entity::Kind::Wolf, { 21, 21 });
	}

	Entity::Ptr World::getEntityAt(unsigned int x, unsigned int y)
	{
		auto found = std::find_if(m_entities.begin(), m_entities.end(),
			[x, y](const Entity::Ptr e) -> bool {
			return e->getPosition().x == x && e->getPosition().y == y;
		});

		return (found != m_entities.end() ? *found : nullptr);
	}

	void World::removeKilledEntites()
	{
		m_entities.erase(
			std::remove_if(
				m_entities.begin(),
				m_entities.end(),
				[](const Entity::Ptr e) -> bool {
					return !e->isAlive();
				}
			),
			m_entities.end()
		);
	}

	void World::sortEntities()
	{
		std::sort(m_entities.begin(), m_entities.end(),
			[](const Entity::Ptr a, const Entity::Ptr b) -> bool {

			unsigned int aInit = a->getInitiative(), bInit = b->getInitiative();

			if (aInit != bInit)
				return a->getInitiative() < b->getInitiative();

			return a->getLifeTime() < b->getLifeTime();
		});
	}

	void World::spawnEntity(Entity::Kind kind, Point position)
	{
		switch (kind) {
		case Entity::Kind::Antelope:
			m_entities.push_back(std::make_shared<Antelope>(*this, position));
			break;
		case Entity::Kind::CyberSheep:
			m_entities.push_back(std::make_shared <CyberSheep>(*this, position));
			break;
		case Entity::Kind::Fox:
			m_entities.push_back(std::make_shared<Fox>(*this, position));
			break;
		case Entity::Kind::Human:
			m_entities.push_back(std::make_shared<Human>(*this, position));
			break;
		case Entity::Kind::Sheep:
			m_entities.push_back(std::make_shared<Sheep>(*this, position));
			break;
		case Entity::Kind::Turtle:
			m_entities.push_back(std::make_shared<Turtle>(*this, position));
			break;
		case Entity::Kind::Wolf:
			m_entities.push_back(std::make_shared<Wolf>(*this, position));
			break;
		case Entity::Kind::Bellandona:
			m_entities.push_back(std::make_shared<Belladona>(*this, position));
			break;
		case Entity::Kind::Dandelion:
			m_entities.push_back(std::make_shared<Dandelion>(*this, position));
			break;
		case Entity::Kind::Grass:
			m_entities.push_back(std::make_shared<Grass>(*this, position));
			break;
		case Entity::Kind::Guarana:
			m_entities.push_back(std::make_shared<Guarana>(*this, position));
			break;
		case Entity::Kind::Hogweed:
			m_entities.push_back(std::make_shared<Hogweed>(*this, position));
			break;
		}
	}

	void World::tick()
	{
		removeKilledEntites();
		sortEntities();

		std::size_t size = m_entities.size();
		for (std::size_t i = 0; i < size; ++i)
			if (m_entities[i]->isAlive()) {
				m_entities[i]->action();
				const Point & position = m_entities[i]->getPosition();
				auto e = getEntityAt(position.x, position.y);
				if (e) {
					m_entities[i]->collision(*e);
					e->collision(*m_entities[i]);
				}
				m_entities[i]->incrementLifeTime();
			}
	}

	void World::printLegend() const
	{
		std::cout << "Plants:           Animals:        Special:\n"
				  << "BD - Belladonna   AT - Antilope   HM - Human\n"
				  << "DN - Dandelion    FX - Fox        CP - CyberSheep\n"
				  << "GS - Grass        SP - Sheep\n"
				  << "GA - Guarana      TT - Turtle\n"
				  << "HW - Hogweed      WF - Wolf";
	}

	void World::printSpacer() const
	{
		std::cout << "\n=================================================\n";
	}

	void World::print()
	{
		std::system("cls");
		std::cout << "Kontanty Misiak 175524\n";
		printSpacer();
		printLegend();
		printSpacer();

		m_area.clear();
		for (const auto & e : m_entities)
			e->draw(m_area);
		m_area.print();
	}

} // namepace bots

File: src/com/kostu96/bots/World.java
package com.kostu96.bots;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Vector;
import java.util.function.Predicate;

import javax.swing.BoxLayout;
import javax.swing.ImageIcon;
import javax.swing.JPanel;

import com.kostu96.bots.entities.Entity;
import com.kostu96.bots.entities.animals.*;
import com.kostu96.bots.entities.plants.*;
import com.kostu96.bots.utils.ImageManager;
import com.kostu96.bots.utils.ImageManager.ImageID;
import com.kostu96.bots.utils.Point;

public class World extends JPanel {
	private static final long serialVersionUID = 1L;
	
	private int columns, rows;
	private List<Entity> entities;
	private List<Cell> tiles;
	private boolean[][] area;
	
	private void addEntity(Entity e) {
		Point p = e.getPosition();
		entities.add(e);
		area[p.x][p.y] = true;
		
		// TODO: messages
		System.out.println(e.getClassName() + " was spawned at " + p);
	}
	
	private void updateImages() {
		for (int i = 0; i < rows; ++i) {
			for (int j = 0; j < columns; ++j) {
				Cell c = getTileAt(j, i);
				Entity e = getEntityAt(j, i);
				try {
					if (e != null) {
						c.setIcon(new ImageIcon(ImageManager.getImage(e.getImageID())));
						c.setToolTipText(e.getClassName());
					} else {
						c.setIcon(new ImageIcon(ImageManager.getImage(ImageID.EMPTY)));
						c.setToolTipText(null);
					}
				} catch (IOException e1) {
					e1.printStackTrace();
				}
			}
		}
	}
	
	private void removeKilledEntities() {
		entities.removeIf(new Predicate<>() {
			@Override
			public boolean test(Entity t) {
				if (!t.isAlive()) {
					Point p = t.getPosition();
					area[p.x][p.y]= false;
					return true;
				}
				return false;
			}
		});
	}
	
	private void sortEntities() {
		entities.sort(new Comparator<Entity>() {
			@Override
			public int compare(Entity o1, Entity o2) {
				if (o1.getInitiative() > o2.getInitiative())
					return -1;
				if (o1.getInitiative() < o2.getInitiative())
					return 1;
				if (o1.getLifeTime() > o2.getLifeTime())
					return -1;
				if (o1.getLifeTime() < o2.getLifeTime())
					return 1;
				return 0;
			}
		});
	}
	
	private Cell getTileAt(int x, int y) {
		return tiles.get(y * columns + x);
	}
	
	public World(int columns, int rows) throws IOException {
		super();
		
		this.columns = columns;
		this.rows = rows;
		
		entities = new ArrayList<>();
		area = new boolean[columns][rows];
		tiles = new ArrayList<>();
		
		addEntity(new Antelope(this, new Point(1, 1)));
		addEntity(new Antelope(this, new Point(1, 3)));
		addEntity(new Fox(this, new Point(4, 5)));
		addEntity(new Fox(this, new Point(6, 5)));
		addEntity(new Sheep(this, new Point(4, 7)));
		addEntity(new Sheep(this, new Point(4, 9)));
		addEntity(new Turtle(this, new Point(5, 9)));
		addEntity(new Turtle(this, new Point(6, 9)));
		addEntity(new Wolf(this, new Point(6, 11)));
		addEntity(new Wolf(this, new Point(6, 13)));
		addEntity(new Belladonna(this, new Point(2, 13)));
		addEntity(new Dandelion(this, new Point(3, 15)));
		addEntity(new Grass(this, new Point(4, 17)));
		addEntity(new Guarana(this, new Point(12, 12)));
		addEntity(new Hogweed(this, new Point(5, 5)));
		//addEntity(new Human(this, new Point(10, 10)));
				
		setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
		for (int i = 0; i < rows; ++i) {
			JPanel row = new JPanel();
			row.setLayout(new BoxLayout(row, BoxLayout.X_AXIS));
			row.setAlignmentX(LEFT_ALIGNMENT);
			for (int j = 0; j < columns; ++j) {
				Cell c;
				if (area[j][i])	{
					Entity e = getEntityAt(j, i);
					c = new Cell(ImageManager.getImage(e.getImageID()));
					c.setToolTipText(e.getClassName());
					row.add(c);
				}
				else {
					c = new Cell(ImageManager.getImage(ImageID.EMPTY));
					row.add(c);
				}
				tiles.add(c);
			}
			add(row);
		}
	}
	
	public int getColumns() { return columns; }
	
	public int getRows() { return rows; }
	
	public void turn() {
		sortEntities();
		
		int size = entities.size();
		for (int i = 0; i < size; ++i) {
			Entity e = entities.get(i);
			if (e.isAlive()) {
				e.action();
				Point p = e.getPosition();
				Entity x = getEntityAt(p.x, p.y);
				if (x != null && x != e)
					x.collision(e);
				e.incrementLifeTime();
			}
		}
		removeKilledEntities();
		updateImages();
	}
	
	public Entity getEntityAt(int x, int y) {
		for(Entity e : entities)
			if(e.getPosition().x == x && e.getPosition().y == y)
		    	return e;
		
		return null;
	}
	
	public Vector<Point> getFreeSpaceAround(Point pos) {
		Vector<Point> arr = new Vector<>();

		int nx = (pos.x - 1 < 0 ? pos.x - 1 + columns : (pos.x - 1) % columns);
		if (!area[nx][pos.y])
			arr.add(new Point(nx, pos.y));

		nx = (pos.x + 1 < 0 ? pos.x + 1 + columns : (pos.x + 1) % columns);
		if (!area[nx][pos.y])
			arr.add(new Point(nx, pos.y));

		int ny = (pos.y - 1 < 0 ? pos.y - 1 + rows : (pos.y - 1) % rows);
		if (!area[pos.x][ny])
			arr.add(new Point(pos.x, ny));

		ny = (pos.y + 1 < 0 ? pos.y + 1 + rows : (pos.y + 1) % rows);
		if (!area[pos.x][ny])
			arr.add(new Point(pos.x, ny));

		return arr; 
	}
	
	public void spawnEntity(String className, Point position) {
		switch (className) {
		case "Antelope": addEntity(new Antelope(this, position)); break;
		case "CyberSheep": addEntity(new CyberSheep(this, position)); break;
		case "Fox": addEntity(new Fox(this, position)); break;
		case "Human": addEntity(new Human(this, position)); break;
		case "Sheep": addEntity(new Sheep(this, position)); break;
		case "Turtle": addEntity(new Turtle(this, position)); break;
		case "Wolf": addEntity( new Wolf(this, position)); break;
		case "Belladonna": addEntity(new Belladonna(this, position)); break;
		case "Dandelion": addEntity(new Dandelion(this, position)); break;
		case "Grass": addEntity(new Grass(this, position)); break;
		case "Guarana": addEntity(new Guarana(this, position)); break;
		case "Hogweed": addEntity(new Hogweed(this, position)); break;
		default: break;
		}
	}
}
